Fix x86 stackwalk for UnmanagedCallersOnly

janvorli opened the pull request:
Yet another similar issue to the recently fixed x86 EH for methods
marked with UnmanagedCallersOnly, but called from a managed method using
a function pointer. The stack walker was asserting in
VerifyValidTransitionFromManagedCode. The REGDISPLAY values that it used
for checking the code location were not set for this case, so EIP and ESP
were both set to 0.
This change fixes it by adding UpdateRegDisplay method to
UMThkCallFrame.

Close #60648
Close https://github.com/dotnet/runtime/issues/62728

janvorli committed:
Fix x86 stackwalk for UnmanagedCallersOnly

Yet another similar issue to the recently fixed x86 EH for methods
marked with UnmanagedCallersOnly, but called from a managed method using
a function pointer. The stack walker was asserting in
VerifyValidTransitionFromManagedCode. The REGDISPLAY values that it used
for checking the code location were not set for this case, so EIP and ESP
were both set to 0.
This change fixes it by adding UpdateRegDisplay method to
UMThkCallFrame.

jkotas commented:
Is this a problem from `class ComMethodFrame` as well?

janvorli commented:
If it is possible to call an UnmanagedCallersOnly method via COM from a managed method so that there are no intermediate native frames, then it would probably suffer from the same issue. @AaronRobinsonMSFT can such a case exist?

AaronRobinsonMSFT commented:
> If it is possible to call an UnmanagedCallersOnly method via COM from a managed method so that there are no intermediate native frames, then it would probably suffer from the same issue. @AaronRobinsonMSFT can such a case exist?

@janvorli How about an example of what I think you are saying. Is the following the question? If so, this is completely reasonable code.

```csharp
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

unsafe class Program
{
    struct Vtable
    {
        public delegate* unmanaged<Inst*, int> Method1;
    }
    
    struct Inst
    {
        public Vtable* vptr;
    }
    
    [UnmanagedCallersOnly]
    static int MethodImpl(Inst* p) => throw new Exception();
    
    static void Main()
    {
        var vptr = (Vtable*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(Program), sizeof(Vtable));
        vptr->Method1 = &MethodImpl;
        var inst = new Inst() {vptr = vptr};
        
        // Imagine inst was ambiguous - could be a real COM object
        // or a managed COM wrapper we don't know about.
        inst.vptr->Method1(&inst);
    }
}
```

janvorli commented:
@AaronRobinsonMSFT my question was about real COM calls that use ComMethodFrame. I assume your example doesn't end up using that frame, does it?
I'll try to rephrase my question a bit. The issue that I am fixing here occurs only when a managed method directly calls another managed method marked with UnmanagedCallersOnly and it does that via a function pointer. Then there is no native frame between the caller and callee frames. @jkotas asked me if it is possible that the same case (no native frame between the caller and callee frames) happens when a managed method directly calls another managed method via COM. That would mean that the COMMethodFrame would be created inside of the managed caller.

jkotas commented:
Hmmm, I am actually wondering how the fix in `UMThkCallFrame` is related to `UnmanagedCallersOnly` methods. `UMThkCallFrame` should not be involved at all in `UnmanagedCallersOnly` methods. `UMThkCallFrame` should be only involved in marshaled reverse-PInvoke delegates.

What is the exact set of conditions that triggers the bug?

jkotas commented:
Here is the callstack that triggers the problem:
```
...
coreclr!WKS::GCHeap::WaitUntilGCComplete+0x4f [D:\a\_work\1\s\src\coreclr\gc\gcee.cpp @ 287] 
...
coreclr!JitILStub+0x9e [D:\a\_work\1\s\src\coreclr\vm\dllimport.cpp @ 5598] 
coreclr!UMThunkMarshInfo::RunTimeInit+0xd2 [D:\a\_work\1\s\src\coreclr\vm\dllimportcallback.cpp @ 480] 
coreclr!UMEntryThunk::RunTimeInit+0xab [D:\a\_work\1\s\src\coreclr\vm\dllimportcallback.h @ 182] 
coreclr!UMEntryThunk::DoRunTimeInit+0xde [D:\a\_work\1\s\src\coreclr\vm\dllimportcallback.cpp @ 282] 
0x3233087 // UMEntryPrestub
Microsoft_Interop_Tests_NativeExports!NativeExports.DelegatesAndFunctionPointers.InvokeCallbackAfterGCCollect(Void ())+0x6d
...
```
As expected, there is no `UnmanagedCallersOnly` method. It is a prestub for marshaled delegate that triggers the problem.

I think we will get into the same situation with COM. `UMEntryPrestub` is going to be `GenericComCallStub`, `UMThkCallFrame` is going to be `ComMethodFrame`, and the rest of the structure is going to be similar.

janvorli commented:
I am sorry for confusing the explanation. But I believe the `UnmanagedCallersOnly` is what's triggering the problem. The `InvokeCallbackAfterGCCollect` is marked with this attribute and the method it calls is invoked via an unmanaged function pointer. When the function pointer is called, the UMEntryPrestub is invoked, the UMTkhCallFrame is created in that prestub and we go down the way of jitting a related IL stub. 
That's the point where the stack walk walks the stack and hits the problem. It gets to the UMTkhCallFrame and then unwinds to the caller, which is a managed method. The fact that it is a managed method triggers the issue.

janvorli commented:
@jkotas actually, let me correct myself. After sorting this in my head again, I have found I've confused it in my previous comment and I was right in my original description. The issue is really caused by the fact that we call a managed function marked using the `UnmanagedCallersOnly` via a native function pointer. If the target of the pointer was a native method, there would be no `UMThkCallFrame`. And it doesn't matter whether the managed caller of the function pointer is marked with `UnmanagedCallersOnly` or not.
But I think you are right with the ComMethodFrame having the same problem.

jkotas commented on src/coreclr/vm/i386/cgenx86.cpp:
```diff
@@ -200,6 +200,41 @@ void EHContext::UpdateFrame(PREGDISPLAY regs)
+#undef CALLEE_SAVED_REGISTER
+
+    pRD->ControlPC = *PTR_PCODE(pRD->PCTAddr);
+    UINT cbStackPop = GetMethod()->CbStackPop();
```
If I am reading the code correctly, `GetMethod` is going to return the managed method that the UMThunk or ComMethodFrame is wrapping.

Is it the managed method to ask for `CbStackPop`? I suspect we need to `CbStackPop` for the unmanaged signature with unmanaged view for the arguments here.

janvorli replied:
Hmm, yes, you are right, I have forgotten that the unmanaged view can be different.

AaronRobinsonMSFT replied:
@janvorli Do we have a test for this? If you can offer a simple strawman example the Interop team can help with adding the appropriate testing assets.

janvorli replied:
@AaronRobinsonMSFT to test this fully, we would need:
* A COM test that exposes a managed method via COM and calls it from other managed method
* That method would need to have arguments passed on stack
* The unmanaged signature would need to differ from the managed one in the size of arguments passed on stack.

It would be great if someone from the interop team can help me with creating that.

janvorli replied:
@AaronRobinsonMSFT I also have a question about getting the unmanaged signature for COM method. Would the right one be the one returned by the `ComCallMethodDesc::GetCallMethodDesc`?

AaronRobinsonMSFT commented on src/coreclr/vm/i386/cgenx86.cpp:
```diff
@@ -200,6 +200,41 @@ void EHContext::UpdateFrame(PREGDISPLAY regs)
+    }
+    CONTRACT_END;
+
+    pRD->PCTAddr = GetReturnAddressPtr();
```
This seems odd. The input is marked `const` but we are updating it?

janvorli replied:
In this case the const-ness is on the pointer itself, not on the value it points to.  I'm not sure why the const is on the parameter of the function (it is an override of a base virtual one that has it that way), maybe it is called on a const Frame derived object somewhere.
The PREGDISPLAY is a typedef `typedef REGDISPLAY *PREGDISPLAY` and that causes the `const PREGDISPLAY` to apply the const-ness  to the pointer. In other words, it is as if the argument was `REGDISPLAY * const pRD`.

AaronRobinsonMSFT replied:
@janvorli I get the C++ semantics here, that isn't my concern. It is the intent. My guess here is the original author didn't realize the `typedef` nuance here so I am simply trying to reconcile what the `const` should mean rather than what it semantically does in the language. Should we change it?

janvorli replied:
To make sure I understand, you'd prefer the `const` to be removed to not to confuse people into thinking the contents of the struct is const?

AaronRobinsonMSFT replied:
Ideally, yes. However, it was merely an observation. Once I checked saw the `typedef` I knew exactly what was going on, but it is one of those odd things I've notice people can trip up on. If the `const` isn't really doing anything, I think it would be good to remove. That is simply quality of life stuff though and if it is more than one or two places I'd leave it.

janvorli replied:
It is at 54 places in 12 files

AaronRobinsonMSFT replied:
> It is at 54 places in 12 files

Definitely not worth it.

mangod9 commented:
@janvorli is this ok to merge?

janvorli commented:
@mangod9 no, as @jkotas pointed out above, the MethodDescs that I was using for the signatures need to reflect the caller ones, not the callee ones that I am using there now. I need to figure out where to get the proper ones yet.

AaronRobinsonMSFT commented:
@mangod9 This is sort of blocked on me. I am investigating now so I can provide the clarity for @janvorli.

mangod9 commented:
ok, thanks for the update.

janvorli commented:
@jkotas @AaronRobinsonMSFT helped me to look more into the `ComMethodFrame`. There is only a narrow case of COM implementation where this frame is used and we have found that it cannot occur in case the target is marked with `UnmanagedCallersOnly`. So I am going to revert the `ComMethodFrame` related part.

janvorli committed:
Fix x86 Linux build

janvorli committed:
Let the new UpdateRegDisplay handle only the problematic case

It would be difficult and useless to attempt to correctly handle the
case when the caller is a native code. Getting the stack arguments
size for the caller seems to be difficult, since in some cases,
we reach the UpdateRegDisplay before the reverse pinvoke IL stub and
even its MethodDesc are created.

The case that we care about is simpler, as the caller side and callee
side signatures are the same, so we can extract the stack arguments size
using the callee MethodDesc.

janvorli commented:
@jkotas I was trying to find a way to get the callee side signature for getting the correct stack arguments size. I have realized that for our problematic case, the caller and callee view of the signatures are the same and we don't need to use the UpdateRegDisplay for the other cases. It would be difficult to attempt to correctly handle the case when the caller 
is a native code. Getting the stack arguments size for the caller in that case seems to be difficult, since in some cases,
we reach the UpdateRegDisplay before the reverse pinvoke IL stub and even its MethodDesc are created. And the callee 
side MethodDesc is determined by quite an involved code.

So I have modified my change to let the UpdateRegDisplay do its work only for the case when the caller is a managed method.

jkotas commented:
> the caller and callee view of the signatures are the same

It is just a pure accident that the callback has `delegate void VoidVoid()` signature and that the managed and unmanaged CbStackPop is zero for both cases.

If the signature of the callback was `delegate void VoidInt(int x)`, the managed and unmanaged CbStackPop values would not match. The `CbStackPop` computed by the code in this PR would be `0`, but the correct `CbStackPop` to use would be 4 - assuming the callback has stdcall calling convention.

Also, I am puzzled by the check for managed caller. 

> Getting the stack arguments size for the caller in that case seems to be difficult, since in some cases,
we reach the UpdateRegDisplay before the reverse pinvoke IL stub

Yes, I agree. I think we would have to compute ahead of time in `LoadTimeInit` to make this work.

janvorli commented:
> If the signature of the callback was `delegate void VoidInt(int x)`, the managed and unmanaged CbStackPop values would not match. The `CbStackPop` computed by the code in this PR would be `0`, but the correct `CbStackPop` to use would be 4 - assuming the callback has stdcall calling convention.

When a managed caller calls a managed method marked by the `UnmanagedCallersOnly` attribute using an unmanaged function pointer, I don't see how the signatures  and calling conventions could be different. 

> Also, I am puzzled by the check for managed caller.

That's the only case when the problem can occur. The stack walker walks to the `UMThkCallFrame` and before my change, it worked only if the caller was a native code in such a case, so the `ProcessIp(adr)` doesn't end up setting `m_crawl.isFrameless`.

jkotas commented:
> When a managed caller calls a managed method marked by the UnmanagedCallersOnly attribute using an unmanaged function pointer, I don't see how the signatures and calling conventions could be different.

MethodDesc::CbStackPop does not return the correct amount for methods marked with `UnmanagedCallersOnly` attribute. MethodDesc::CbStackPop assumes managed calling convention. It does not look at the method calling convention.

jkotas commented:
> Can this problem be fixed by deleting UMThkCallFrame?

https://github.com/dotnet/runtime/pull/63826 . Let's see what the CI says about it.

janvorli commented:
That's great! The CI seems to be happy with it.

jkotas commented:
Replaced by #63826
